### ISSUE TITLE
PyInstaller 5.0 can't find icon when running spec file out of cwd

### ISSUE THREAD

**Issue by Numerlor** [pyinstaller/pyinstaller#6759]
## Description of the issue
```
root/
├── file.py
├── icon
│   └── icons_library.ico
├── pyinstaller_build
│   └── file.spec
```
With the above file structure, PyInstaller now tries to resolve the icon both relative to the spec file's location, and the cwd, making the icon to be impossible to find. This worked fine in previous version.

Running from the `root` directory,
with the icon set to `../icon/icons_library.ico` an exception is raised from https://github.com/pyinstaller/pyinstaller/blob/68a79e1c089aece4bfe93eae58b3edaa0ff223ec/PyInstaller/building/icon.py#L32-L34 and printing the path reveals it's relative to the cwd (goes above the root).

When changed to `./icon/icons_library.ico`, PyInstaller ends with `Unable to open icon file ./icon/icons_library.ico` https://github.com/pyinstaller/pyinstaller/blob/68a79e1c089aece4bfe93eae58b3edaa0ff223ec/PyInstaller/utils/win32/icon.py#L100-L106

### Context information (for bug reports)

* Output of `pyinstaller --version`: `5.0`
* Version of Python: 3.10.4
* Platform: Windows
* How you installed Python: python.org/downloads
* Did you also try this on another platform? Does it work there? Tried windows only

### A minimal example program which shows the error
https://github.com/Numerlor/pyinstaller_icon
set up project (install pyinstaller/`poetry install`) and run `pyinstaller .\pyinstaller_build\file.spec` from the project root

**Comment by bwoodsend** [comment]
> With the above file structure, PyInstaller now tries to resolve the icon both relative to the spec file's location, and the cwd, making the icon to be impossible to find. This worked fine in previous version.

I suppose that we need to choose which behaviour we actually want. I vote for `./foo` being equivalent to `$PWD/foo` only and we ditch this *relative to the .spec file* behaviour everywhere it has been inconsistently applied. But whichever we go for - it must be one and not both. If we do choose to anchor paths relative to the spec file then we should apply that to `datas` and `binaries` too and implement it the same way that Sphinx does with its `conf.py` - by `cd`-ing to that file's parent directory before running it.

**Comment by Numerlor** [comment]
> I suppose that we need to choose which behaviour we actually want. I vote for `./foo` being equivalent to `$PWD/foo` only and we ditch this _relative to the .spec file_ behaviour everywhere it has been inconsistently applied.  But whichever we go for - it must be one and not both. If we do choose to anchor paths relative to the spec file then we should apply that to `datas` and `binaries` too ...

Being relative to the spec file is how the paths act now as far as I can tell (at least for datas) as I used paths relative to the spec file in the project where I encountered this bug https://github.com/Numerlor/Auto_Neutron/blob/f17fae47e6b5b92b89f01a274e367b9275c58400/pyinstaller_build/spec_helper.py#L7-L11 

Being relative to the working directory feels like what I'd expect initially, though if it's a file I think there's some merit to the paths inside of it being relative to the file's location (except for pathex, maybe hooks too though I didn't check those) - which is the behaviour now (or up to now).
I'd be vary of changing it all to be relative to the cwd as it is a breaking change, though I'm not sure in how many cases the file's location and working directory differ.

**Comment by The-Compiler** [comment]
FWIW I was hit by this as well, as I have the `.spec` file in a `misc/` subdirectory, but run it from the project root. I assume that's something a few larger projects are doing, but I might be wrong.

As a workaround, an absolute icon path can be constructed in the `.spec` file. Unfortunately, `__file__` doesn't seem to be set, but at least for my case, it was fine to assume that the cwd is the repository root: https://github.com/qutebrowser/qutebrowser/commit/07ce0349837fffc221184857be56517a2d208bd8

**Comment by bwoodsend** [comment]
There's a `SPECPATH` global which points to the parent directory of the spec file for anchoring files relative to it (although I personally wish we just used `__file__`).

**Comment by bwoodsend** [comment]
I guess since `datas` works relative to the spec file then the icon should too. I do still think that if we go down this path then PyInstaller should just `cd` to the spec's parent directory so that `datas = [("foo/*.txt", ".")]` and `datas = [(i, ".") for i in glob.glob("foo/*.txt")]` aren't doing two totally different things but that will have to wait until the next major release.

**Comment by The-Compiler** [comment]
> I do still think that if we go down this path then PyInstaller should just `cd` to the spec's parent directory

Note that would also change where e.g. `build/` and `dist/` end up. I think if I start PyInstaller in the repo root, I would expect the output to be there (or if I want to e.g. test something, I might want to cd to a different output directory). In contrast to that, it makes sense to resolve *input* files relative to the spec file (rather than cwd) so that the spec file works no matter where PyInstaller was invoked from. It would be weird if a `mkdir pyinstaller-test; cd pyinstaller-test; pyinstaller ../myapp.spec` required changes to the spec file, and it wouild also be weird if the output landed in `..`.

**Comment by bwoodsend** [comment]
> It would be weird if a mkdir pyinstaller-test; cd pyinstaller-test; pyinstaller ../myapp.spec required changes to the spec file, and it wouild also be weird if the output landed in ...

Personally I'd like `dist` and `build` to be controlled by the spec and therefore would expect the output to land in `..` so that the `cwd` is effectively irrelevant. But yeah, it doesn't make much sense no matter what we do so just not changing things is probably best. If the spec file was a declarative config/toml/yaml file  instead of half a Python script then the _relative paths are relative to the spec_ pattern would feel less of an inconsistency to me but that's never likely to happen either.

**Comment by Numerlor** [comment]
I still encounter an issue with icons after #6775 was merged, the issue now seems to be https://github.com/pyinstaller/pyinstaller/blob/1a7d704ffbabb433007e3ba04750c2f13ade48e5/PyInstaller/utils/win32/icon.py#L98-L99 joining the path with specpath while it's already joined in the code added in the PR

**Comment by bwoodsend** [comment]
What do you mean by _it's already joined in the code_? Can you open a new issue with a reproducer so that we're clear what the problem is.
